### ISSUE TITLE
Feature request: integrate regex-based extraction mechanisms

### ISSUE THREAD

**Issue by aborruso** [wezm/rsspls#46]
Hi,
in this site with not very good HTML, I use the awesome rsspls to generate an RSS feed:
https://bagheria.trasparenza-valutazione-merito.it/web/trasparenza/papca-ap/-/papca/igrid/39908/24264

I use these settings

```yml
[rsspls]
output = "../docs/c_a546"
[[feed]]
title = "Albo del comune di Bagheria"
filename = "feed.xml"
[feed.config]
url = "https://bagheria.trasparenza-valutazione-merito.it/web/trasparenza/papca-ap/-/papca/igrid/39908/24264"
item = "tbody tr:nth-child(n+2)"
heading = "td.oggetto.text"
link = "a[title=\"Apri Dettaglio\"]"
```

I would also like to extract the publication date which is the first date that appears in the td tag with class “date”

```
<td tabindex="0" role="cell" class="periodo-pubblicazione date">31/05/2024<br>
  15/06/2024</td>
```

But via CSS selector I don't think there is any way to select the first one.

So if I apply the CSS selector, and set `format = "[day padding:none]/[month padding:none]/[year]"` I get these errors, because I can't find a way to tell to disregard the second date as well.

```
 WARN  rsspls::feed > unable to parse date '31/05/2024  15/06/2024'
 WARN  rsspls::feed > unable to parse date '31/05/2024  15/06/2024'
 WARN  rsspls::feed > unable to parse date '31/05/2024  15/06/2024'
 WARN  rsspls::feed > unable to parse date '31/05/2024  15/06/2024'
...
```

It would be great to be able to set some magic char or regex, to escape some text. In example

```
regex="^(.+[0-9]) .+$"
format = "[day padding:none]/[month padding:none]/[year]"
```

to extract only `'31/05/2024` from `31/05/2024  15/06/2024`, and then apply the format.

![image](https://github.com/wezm/rsspls/assets/30607/ddc29b64-c46f-4759-b3c8-aa0bece63722)

Thank you

**Comment by wezm** [comment]
Hmm that is a tricky one. I will admit that I'm feeling a bit hesitant to throw regexes into the mix. In your specific case it seems the page is using a fixed width format for the dates so I think you can use `ignore`:

<img width="713" alt="Screenshot from 2024-06-03 08-58-59" src="https://github.com/wezm/rsspls/assets/21787/ff126ef8-b1f1-4687-94f3-b153712ea424">

This works in my testing:

```toml
format = "[day padding:none]/[month padding:none]/[year][ignore count:12]"
```

However it does feel a bit fragile. A nicer option would be to just stop parsing the date after the year. [time](https://time-rs.github.io/book/api/format-description.html) already has an `end` component:

<img width="742" alt="Screenshot from 2024-06-03 09-00-57" src="https://github.com/wezm/rsspls/assets/21787/0938a519-2db7-4bf4-aa50-1fc6c207acdf">

but it only succeeds if there is no further input after it. I wonder if they would be open to an option on that to allowed terminating parsing when it's encountered, even if there is input remaining.

**Comment by aborruso** [comment]
Thank you @wezm I have a lot of difficulty understanding that documentation.
And I didn't realize, for example, that it was possible to use this

```
format = "[day padding:none]/[month padding:none]/[year][ignore count:12]"
```

It may be fragile, but it seems stable for that site.

But I want to try again to propose the introduction of regexes, as an option for all fields to be extracted: 
- so many pages have bad HTML; 
- css selector alone may not be enough; 
- regular expressions are a standard; 
- you could optionally introduce them after extraction by CSS selector.

```mermaid
graph TD
    A[Webpage] --> B[CSS Selector]
    B --> C{Regex}
    C -- Filter --> D[RSS]
    C -- No filter --> D[RSS]
    style C stroke-dasharray: 5, 5;
```

Unfortunately I am not a developer and I can't help you with the code, however I think it could be a very convenient feature and maybe it is quite straightforward to introduce.

Once again thank you very much.

**Comment by wezm** [comment]
> Thank you @wezm I have a lot of difficulty understanding that documentation.

That feels like something I can try to improve too. Was it the `rsspls` documentation or the `time` library's date parsing documentation that tripped you up (or both)?

> But I want to try again to propose the introduction of regexes, as an option for all fields to be extracted:

I will keep it in mind but I'm not ready to add it yet. Regular expressions are a big hammer to solve a problem with so before incorporating them I'd like to have a decent collection of real-world examples of where they would help solve a problem. I'll leave this issue open to see if such cases come up.

**Comment by aborruso** [comment]
> That feels like something I can try to improve too. Was it the `rsspls` documentation or the `time` library's date parsing documentation that tripped you up (or both)?

I like the documentation of `rsspls`. The one from `time`, it seems to be written for those who already know how it works. It would take examples, first the simplest and most complex (if for example this ... then you have to ...).

> Regular expressions are a big hammer 

Ma non puoi fare come con `time`? Choose a regex Rust library and tell users to refer to it?
And enable only simply regex match. For my example `^(\d{2}/\d{2}/\d{4})`.

Thank you again

**Comment by wezm** [comment]
> I like the documentation of `rsspls`. The one from `time`, it seems to be written for those who already know how it works. It would take examples, first the simplest and most complex (if for example this ... then you have to ...).

Yes the time docs I'm linking to are the developer oriented docs. I think we can do better for end-users.

> But you can't do it like with `time`? Choose a regex Rust library and tell users to refer to it?

Yes, I'm not saying it's hard to do, I'm saying that I'm not certain it's how I want to solve the problem. I'd like to gather more data points before committing to it.

**Comment by P1n3appl3** [comment]
I had the exact same problem as OP and I'd read over the `time` docs but I missed ignore (thx for the pointer Wesley). Unfortunately I think it doesn't solve my use case as I don't have a constant byte count of ignored bytes. The html I'm trying to wrangle looks like:
```html
<a href="https://petrichorstreet.neocities.org/posts/2026-01-20-selfishness-and-reclamation.html">
 2026-01-20 » selfishness and reclamation
</a>
<a href="https://petrichorstreet.neocities.org/posts/2025-12-31-diverging-from-myself.html">
 2025-12-31 » diverging from myself
</a>
<a href="https://petrichorstreet.neocities.org/posts/2025-12-30-performativity-and-the-need-to-be-seen.html">
 2025-12-30 » performativity and the need to be seen
</a>
<a href="https://petrichorstreet.neocities.org/posts/2025-12-23-on-attention-devotion-and-love.html">
 2025-12-23 » on attention devotion and love
</a>
<a href="https://petrichorstreet.neocities.org/posts/2025-11-13-claw-marks.html">
 2025-11-13 » claw marks
</a>
<a href="https://petrichorstreet.neocities.org/posts/2024-12-31-continue.html">
 2024-12-31 » continue
</a>
<a href="https://petrichorstreet.neocities.org/posts/2023-12-31-journey.html">
 2023-12-31 » journey
</a>
```
I thought I'd need regex based extraction to get the date and header for each post, but then your workaround gave me hope, but sadly `format = "[year padding:none]-[month padding:none]-[day] » [ignore count:9999]"` doesn't work since it needs the exact number of bytes to be specified. Your proposed change (time-rs/time#684) would totally fix this issue, and I might just go write that patch for personal use.

The only annoying thing is that even if I got that working for date extraction, the items in the feed include the date in their title too... but honestly I agree with you about being hesitant to add a bunch of functionality for that, and I'm fine with slightly messy titles for a couple blogs in the meantime, so I'd respect the decision to not pull in a regex library and mess with the extractor rules.

**Comment by P1n3appl3** [comment]
I did end up writing a [little patch](https://github.com/wezm/rsspls/commit/d20fa5d0f79df5415abb4f7efb5d981c7b6d0994) to allow regex date extraction.

**Comment by wezm** [comment]
Thanks for your experience report @P1n3appl3. Situations with titles like "2024-12-31 » continue" really are asking for a way to filter the value. It is probably time to add that in some form.

**Comment by P1n3appl3** [comment]
My ideal interface would be that each selector can optionally be a selector + regex at the end to extract the specific text for title/summary/date, something like `header = "div.posts a | '(.+) - '"`. If that's too implicit for your tastes, maybe this would be better: `header = { selector = "div.posts a", regex = "(.+) - " }`. I'm not sure if toml + serde would let you still do `header = "selector"` in that case.

**Comment by wezm** [comment]
> My ideal interface would be that each selector can optionally be a selector + regex at the end to extract the specific text for title/summary/date

Yes I was thinking something along these lines too.

> I'm not sure if toml + serde would let you still do `header = "selector"` in that case.

It's possible, but requires a bit more code. I already do it for `date`, which can be a string or table.

https://github.com/wezm/rsspls/blob/6e746850d0e8ed64d9ca11373ff7f51ddb9d609f/src/config.rs#L184
